File: go/native/native_test.go
package native

import (
	"strings"
	"testing"
)

func TestSignRequestRoundTrip(t *testing.T) {
	n := New("app_test", "secret_123", "https://api.example.com")
	res, err := n.SignRequest("dev_001", `{"a":1}`)
	if err != nil {
		t.Fatalf("SignRequest: %v", err)
	}
	if len(res.Sign) != 64 {
		t.Errorf("sign 应为 64 hex，实际 %d", len(res.Sign))
	}
	if res.Timestamp == "" || len(res.Nonce) < 16 {
		t.Errorf("timestamp/nonce 缺失: %+v", res)
	}
	// 重放验证：同一签名数据应得到相同签名（供服务端复算）
	res2, err := signRequest("secret_123", "dev_001", `{"a":1}`)
	if err != nil || res2.Sign != res.Sign {
		t.Error("同参数签名应一致")
	}
}

func TestVerifyUpgrade(t *testing.T) {
	n := New("app_test", "secret_123", "")
	manifest := `{"has_update":true,"latest_version":"1.0.0"}`
	// 先伪造错误签名
	if ok, _ := n.VerifyUpgrade(manifest, "deadbeef"); ok {
		t.Error("错误签名应校验失败")
	}
	sig := hmacSign("secret_123", manifest)
	if ok, _ := n.VerifyUpgrade(manifest, sig); !ok {
		t.Error("正确签名应校验通过")
	}
}

func TestMachineFingerprintDeterministic(t *testing.T) {
	a, err := machineFingerprint()
	if err != nil {
		t.Fatalf("machineFingerprint: %v", err)
	}
	b, _ := machineFingerprint()
	if a.Hash == "" || a.Hash != b.Hash {
		t.Error("指纹应确定且非空")
	}
	if len(a.Fields) == 0 {
		t.Error("指纹字段不应为空")
	}
}

func TestDetectRisksNoPanic(t *testing.T) {
	res, err := detectRisks()
	if err != nil {
		t.Fatalf("detectRisks: %v", err)
	}
	for _, f := range res.Flags {
		switch f {
		case "debug", "emulator", "root", "hook", "multiopen":
		default:
			t.Errorf("未知风险 flag: %s", f)
		}
	}
}

func TestSelfIntegrity(t *testing.T) {
	res, err := executableHash()
	if err != nil {
		t.Fatalf("executableHash: %v", err)
	}
	if len(res) != 32 {
		t.Errorf("SHA256 长度应为 32，实际 %d", len(res))
	}
}

func TestAcquireMutexTwice(t *testing.T) {
	first, err := acquireMutex("test_app_mutex")
	if err != nil {
		t.Fatalf("acquireMutex: %v", err)
	}
	if !first {
		t.Skip("环境已存在实例锁")
	}
	second, _ := acquireMutex("test_app_mutex")
	if second {
		t.Error("同一进程第二次获取应返回 false")
	}
}

func TestSecureStoreRoundTrip(t *testing.T) {
	key := "test_token_1"
	_ = secureSet(key, "secret-value-xyz")
	got, err := secureGet(key)
	if err != nil {
		t.Fatalf("secureGet: %v", err)
	}
	if got != "secret-value-xyz" {
		t.Errorf("读取值不一致: %q", got)
	}
}

func TestSignDataFormat(t *testing.T) {
	// 与平台 DeviceSign 协议对齐：deviceID\n timestamp \n body
	sig, err := signRequest("k", "dev", "body")
	if err != nil {
		t.Fatal(err)
	}
	// 直接按协议构造签名数据复算
	signData := "dev\n" + sig.Timestamp + "\nbody"
	if !strings.EqualFold(hmacSign("k", signData), sig.Sign) {
		t.Error("签名协议与平台不一致")
	}
}

File: rust/native/tests/native_test.rs
use aohoyo_native::Native;

#[test]
fn sign_request_round_trip() {
    let n = Native::new("app_test", "secret_123", "https://api.example.com");
    let res = n.sign_request("dev_001", r#"{"a":1}"#).expect("sign");
    assert_eq!(res.sign.len(), 64, "sign 应为 64 hex");
    assert!(!res.timestamp.is_empty());
    assert!(res.nonce.len() >= 16);

    // 同参数签名应一致（服务端可复算）
    let res2 = n.sign_request("dev_001", r#"{"a":1}"#).expect("sign2");
    assert_eq!(res2.sign, res.sign);
}

#[test]
fn verify_upgrade_signature() {
    let n = Native::new("app_test", "secret_123", "");
    let manifest = r#"{"has_update":true,"latest_version":"1.0.0"}"#;
    assert!(!n.verify_upgrade(manifest, "deadbeef"), "错误签名应失败");

    // 服务端同构：HMAC-SHA256(app_secret, manifestJSON)
    let sig = {
        use hmac::{Hmac, Mac};
        use sha2::Sha256;
        let mut mac = Hmac::<Sha256>::new_from_slice(b"secret_123").unwrap();
        mac.update(manifest.as_bytes());
        hex::encode(mac.finalize().into_bytes())
    };
    assert!(n.verify_upgrade(manifest, &sig), "正确签名应通过");
}

#[test]
fn machine_fingerprint_deterministic() {
    let n = Native::new("app_test", "secret_123", "");
    let a = n.machine_fingerprint().expect("fp1");
    let b = n.machine_fingerprint().expect("fp2");
    assert!(!a.hash.is_empty());
    assert_eq!(a.hash, b.hash, "指纹应确定");
}

#[test]
fn detect_risks_valid_flags() {
    let n = Native::new("app_test", "secret_123", "");
    let res = n.detect_risks().expect("risks");
    for f in &res.flags {
        assert!(
            matches!(
                f.as_str(),
                "debug" | "emulator" | "root" | "hook" | "multiopen"
            ),
            "未知 flag: {f}"
        );
    }
}

#[test]
fn self_integrity_returns_hash() {
    let n = Native::new("app_test", "secret_123", "");
    let res = n.self_integrity().expect("integrity");
    assert_eq!(res.hash.len(), 64, "SHA256 hex 长度应为 64");
    assert!(res.ok);
}

#[test]
fn acquire_mutex_twice() {
    let n = Native::new("app_test", "secret_123", "");
    let first = n.acquire_mutex().expect("mutex1");
    if !first {
        return; // 环境已有实例锁
    }
    let second = n.acquire_mutex().expect("mutex2");
    assert!(!second, "同一进程第二次获取应返回 false");
}

#[test]
fn secure_store_round_trip() {
    let n = Native::new("app_test", "secret_123", "");
    let key = "test_token_rust";
    n.secure_set(key, "secret-value-xyz").expect("set");
    let got = n.secure_get(key).expect("get");
    assert_eq!(got, "secret-value-xyz");
}

#[test]
fn cert_pin_round_trip() {
    let n = Native::new("app_test", "secret_123", "");
    n.set_cert_pin("sha256/abc123");
    assert_eq!(n.get_cert_pin().as_deref(), Some("sha256/abc123"));
}
